kernel-spark/src/main/java/io/delta/kernel/spark/snapshot/unitycatalog/UCManagedTableSnapshotManager.java

```diff
@@ -55,37 +59,123 @@ public UCManagedTableSnapshotManager(
+   *     available log files; if false, considers all commits
+   * @param canReturnEarliestCommit if true, returns the earliest commit if the timestamp is before
+   *     all commits; if false, throws an exception
+   * @return the commit that was active at the specified timestamp
+   */
   @Override
   public DeltaHistoryManager.Commit getActiveCommitAtTime(
       long timestampMillis,
       boolean canReturnLastCommit,
       boolean mustBeRecreatable,
       boolean canReturnEarliestCommit) {
-    throw new UnsupportedOperationException(
-        "UCManagedTableSnapshotManager.getActiveCommitAtTime is not yet implemented");
+    SnapshotImpl snapshot = (SnapshotImpl) loadLatestSnapshot();
+    List<ParsedCatalogCommitData> catalogCommits = snapshot.getLogSegment().getAllCatalogCommits();
+    return DeltaHistoryManager.getActiveCommitAtTimestamp(
+        engine,
+        snapshot,
+        snapshot.getLogPath(),
+        timestampMillis,
+        mustBeRecreatable,
+        canReturnLastCommit,
+        canReturnEarliestCommit,
+        catalogCommits);
   }
 
+  /**
+   * Checks if a specific version exists and is accessible.
+   *
+   * <p>For UC-managed tables with catalogManaged, log files may be cleaned up, so we need to use
+   * DeltaHistoryManager to find the earliest available version based on filesystem state.
+   *
+   * @param version the version to check
+   * @param mustBeRecreatable if true, requires that the version can be fully recreated from
+   *     available log files; if false, only requires that the version's log file exists
+   * @param allowOutOfRange if true, allows versions greater than the latest version without
+   *     throwing an exception; if false, throws exception for out-of-range versions
+   * @throws VersionNotFoundException if the version is not available or does not meet the specified
+   *     criteria
+   */
   @Override
-  public void checkVersionExists(long version, boolean mustBeRecreatable, boolean allowOutOfRange) {
-    throw new UnsupportedOperationException(
-        "UCManagedTableSnapshotManager.checkVersionExists is not yet implemented");
+  public void checkVersionExists(long version, boolean mustBeRecreatable, boolean allowOutOfRange)
+      throws VersionNotFoundException {
+    // Load latest to get the current version bounds
+    SnapshotImpl snapshot = (SnapshotImpl) loadLatestSnapshot();
+    // Latest version visible in this UC-managed snapshot.
+    long latestSnapshotVersion = snapshot.getVersion();
+
+    // Fast path: check upper bound before expensive filesystem operations
+    if ((version > latestSnapshotVersion) && !allowOutOfRange) {
+      throw new VersionNotFoundException(version, 0 /* earliest */, latestSnapshotVersion);
+    }
+
+    // Compute earliestRatifiedCommitVersion from catalog commits
+    List<ParsedCatalogCommitData> catalogCommits = snapshot.getLogSegment().getAllCatalogCommits();
+    Optional<Long> earliestRatifiedCommitVersion =
```

Comment:
Is the goal here to learn: which commits here are catalog commits and not published?

I think we are getting confused on ratified vs catalog commit. I'd suggest just calling it `earliestCatalogCommit` ... but even then .. unlear if you mean earliestCatalogCommitThatIsNotPublished. I'm not quite sure on the semantic you are aiming for.